examples/kfree_skb.bpf.c

```diff
@@ -1,19 +1,98 @@
 #include <vmlinux.h>
 #include <bpf/bpf_tracing.h>
+#include <bpf/bpf_endian.h>
 #include "maps.bpf.h"
 
+#define ETH_P_IPV6 0x86DD
+#define ETH_P_IP 0x0800
+
+struct kfree_skb_key_t {
+    u16 eth_proto;
+    u16 ip_proto;
+    u16 port;
+    u16 reason;
+};
+
 struct {
     __uint(type, BPF_MAP_TYPE_HASH);
-    __uint(max_entries, 1024);
-    __type(key, u16);
+    __uint(max_entries, 10240);
+    __type(key, struct kfree_skb_key_t);
     __type(value, u64);
 } kfree_skb_total SEC(".maps");
 
-SEC("tp_btf/kfree_skb")
-int BPF_PROG(kfree_skb, struct sk_buff *skb, void *location, enum skb_drop_reason reason)
+SEC("tp_btf/kfree_skb") int BPF_PROG(kfree_skb, struct sk_buff *skb, void *location, enum skb_drop_reason reason)
 {
-    u16 key = reason;
+    struct kfree_skb_key_t key;
+    struct ethhdr eth_hdr;
+    struct iphdr ip_hdr;
+    struct ipv6hdr ipv6_hdr;
+    struct tcphdr tcp_hdr;
+    struct udphdr udp_hdr;
+    u16 ip_proto = 0;
+
+    // Same as skb_mac_header_was_set:
+    // * https://elixir.bootlin.com/linux/v6.5-rc1/source/include/linux/skbuff.h#L2899
+    if (skb->mac_header == (typeof(skb->mac_header)) ~0U) {
+        return 0;
+    }
+
+    if (bpf_probe_read_kernel(&eth_hdr, sizeof(eth_hdr), skb->head + skb->mac_header)) {
+        return 0;
+    }
+
+    key.eth_proto = bpf_ntohs(eth_hdr.h_proto);
+
+    if (!key.eth_proto && !bpf_ntohs(skb->protocol)) {
+        return 0;
+    }
+
+    switch (key.eth_proto) {
```

Comment:
Do we have to be vlan aware here? What about other encapsulation?

See: https://github.com/dsahern/bpf-progs/blob/175e2dfd/src/parse_pkt.c#L167-L179